### ISSUE TITLE
Make sure DPoP is passing with official OIDC testsuite

### ISSUE THREAD

**Issue by mposolda** [keycloak/keycloak#31970]
The OIDF Conformance testing suite just added support for testing DPoP support. Link: https://gitlab.com/openid/conformance-suite/-/releases/release-v5.1.20

It can be good to verify that Keycloak is able to pass with the testsuite before we promote "DPoP" as officially supported feature.

Maybe as a follow-up (or together with this task), this can involve to make sure we can test this automatically (like for instance FAPI and OIDC tests in https://github.com/keycloak/kc-sig-fapi/ ) and then obtain certification for DPoP from OIDF.

**Comment by tnorimat** [comment]
@mposolda @thomasdarimont Hello, I ran conformance tests of FAPI 2.0 Security Profile with DPoP (FAPI2-SP DPoP) and FAPI 2.0 Message Signing with DPoP (FAPI2-MS DPoP) against Keycloak 25 and found the following issue that are not directly related to Keycloak itself:

Not only these conformance tests but other conformance tests of FAPI checks both an authorization server like Keycloak and resource server. Therefore, OAuth SIG's automated conformance test environment implemented [the resource server](https://github.com/keycloak/kc-sig-fapi/blob/main/conformance-tests-env/common/api-gateway-nginx/fapi-verify.lua) by itself.

In these conformance tests run, tests for checking whether DPoP is supported against the resource server failed because the resource server does not support DPoP.

In order for Keycloak to pass these conformance tests, we need to support DPoP to [the resource server](https://github.com/keycloak/kc-sig-fapi/blob/main/conformance-tests-env/common/api-gateway-nginx/fapi-verify.lua).

This resource server uses openresty's library ["openidc"](https://github.com/zmartzone/lua-resty-openidc/blob/master/lib/resty/openidc.lua#L1631) written in Lua and applying Apache License 2.0 to verify an access token, and this library does not suppot an access token in DPoP type.

Just an idea is to modify the library's code locally to support DPoP, but I am not familiar with Lua. Therefore, I am happy if someone who is familiar with Lua could do it.

My question about conformance tests is why the conformance tests checks a resource server. I think that the conformance tests are for an authorization server, not a resource server. @thomasdarimont What do you think about this question?

**Comment by thomasdarimont** [comment]
Thank you Takashi for taking a look!

If openresty currently does not support dpop, we could perhaps switch to https://github.com/OpenIDC/mod_auth_openidc which recently added support for dpop.

We could also file an issue on the openresty repo to ask for dpop implementation.

**Comment by tnorimat** [comment]
@thomasdarimont Thank you for your advice and help. I will check mod_auth_openidc.

**Comment by thomasdarimont** [comment]
You're very welcome! 

Perhaps we can also build dpop support into openresty based on that example: https://github.com/curityio/sender-constrained-token-plugin

**Comment by thomasdarimont** [comment]
As an alternative, we could just extend the token introspection endpoint to support dpop proofs, then the resource server could just call the introspection endpoint.

**Comment by thomasdarimont** [comment]
@tnorimat I got a response from the OpenID Certification team regarding the need for DPoP support on the resource server in the testing environment.
Since the system needs a component that verifies the DPoP proof in the test it's logical to use the resource server for that.
An alternative is to just call the userinfo endpoint to thest DPoP.  To do this, just set the resource endpoint in the test config to the userinfo endpoint url.

**Comment by tnorimat** [comment]
@thomasdarimont Thank you very much for your swift help! I will consider your suggesting several options.

**Comment by VinodAnandan** [comment]
@mposolda @tnorimat @thomasdarimont I hope you're both doing well. Have any of you had a chance to test Keycloak's DPoP implementation with the conformance suite yet? I also think it would be beneficial to test the master branch, as it includes several new contributions related to DPoP from @Captain-P-Goldfish 

Cc: @francis-pouatcha @dteleguin

**Comment by tnorimat** [comment]
@thomasdarimont Hello, yes, I will try to use UserInfo Endpoint of keycloak instead of a resource server. I looked into `UserInfoEndpoint.java` and found that it checks DPoP, so I expect we can use it.

**Comment by tnorimat** [comment]
@thomasdarimont I tried to UserInfo Endpoint of keycloak instead of a resource server, but found two problems as follows:

1. Need to return "Date" http header, but UserInfo Endpoint does not do that ([FAPI1 baseline Final - section 6.2.1 #10](https://openid.net/specs/openid-financial-api-part-1-1_0-final.html#rfc.section.6.2.1)).
2. Need to retrurn "x-fapi-interaction-id", but UserInfo Endpoint does not do that ([FAPI1 baseline Final - section 6.2.1 #11](https://openid.net/specs/openid-financial-api-part-1-1_0-final.html#rfc.section.6.2.1)).

To solve the two problems, I considered to modify a [load balancer (HA proxy)](https://github.com/keycloak/kc-sig-fapi/blob/main/conformance-tests-env/common/load-balancer/haproxy.cfg) in front of Keycloak to do HTTP response rewrite, but it seems difficult because the HA proxy works as L4 load balancer, not L7 load balancer.

**Comment by thomasdarimont** [comment]
Thanks @tnorimat for giving this a try! 

ad 1) I think we could simply add the `Date` header to the UserInfoResponse in general in [UserInfoEndpoint#issueUserInfo](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java#L159C22-L159C35).

ad 2) Regarding the `x-fapi-interaction-id` that's tricky, we could simply send the value of that header back in the response or generate a UUID value and send that back, but we should only do that when we are in a "FAPI" context. Is there a way to detect if the user info / resource request should follow FAPI profile rules? Maybe a scope, or a client policy that we can evaluate?
Perhaps we could evaluate a client policy based on the azp claim of the access token used to call the userinfo endpoint. If the FAPI profile is active for the client, we could handle the `x-fapi-interaction-id` request/response header with a client policy.
Perhaps @mposolda has some thoughts here as well.

In the meantime I think we could work around both lacking features by doing the following:
Proxy Keycloaks UserInfo endpoint:
I saw that there is an [OpenResty based API gateway](https://github.com/keycloak/kc-sig-fapi/tree/main/conformance-tests-env/common/api-gateway-nginx). We could define a "custom user info" resource there that simply proxies the request to Keycloak's userinfo endpoint. Then Keycloaks userinfo endpoint could perform the DPoP check and generate the response (with the lacking information). In the nginx/openresty/lua part you could add the Date header and replay/generate a `x-fapi-interaction-id` header in the response.

In the FAPI conformance test we then use the "custom user info" resource as the resource endpoint in the test.

**Comment by tnorimat** [comment]
@thomasdarimont Thank you for the suggestion. Firstly, I will try to modify as you suggested an [API gateway](https://github.com/keycloak/kc-sig-fapi/blob/main/conformance-tests-env/common/api-gateway-nginx/nginx.conf.template) in front of a resource server in the FAPI conformance test platform. 

As for client policies in UserInfo Endpoint, I am afraid that codes in client policies cannot reach [response builder](https://github.com/keycloak/keycloak/blob/dcf1cb50fc487729385804b437eb6bccd38b010f/services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java#L284) so client policies could not handle response headers.

**Comment by tnorimat** [comment]
@thomasdarimont Hello, it seems that the way suggested in https://github.com/keycloak/keycloak/issues/31970#issuecomment-2434497855 have some problem.

Due to its specification, DPoP needs to check whether the following two matches exactly:
1. "htu" claim value in DPoP Proof that Keycloak's Userinfo Endpoint receives via Nginx API gateway of the conformance test environment of OAuth SIG.  = "https://rs,keycloak-fapi.org/custom-userinfo"
2. Userinfo endpoint's URI exposed by Keycloak. = "https://as,keycloak-fapi.org/custom-userinfo"

As for the latter, in docker-compose.yml of the test environment, Keycloak's "hostname" parameter value is "as.keycloak-fapi.org" because a Load Balancer (HA Proxy) sits in front of Keycloak and a browser and client access Keycloak whose host name is "as.keycloak-fapi.org".

Therefore, I would like to pursuit the way of implementing DPoP verification on a lua script as you suggested in https://github.com/keycloak/keycloak/issues/31970#issuecomment-2275096700 .

**Comment by tnorimat** [comment]
@thomasdarimont @mposolda Hello, I supported some manual FAPI2 DPoP tests by your help. Thank you.
https://github.com/keycloak/kc-sig-fapi/pull/641
However, it is not mature so I would like to refine it.

Also, some tests fail due to Keycloak itself and the resource server I implemented on the test environment.
I try to resolve the failed test due to the resource server, and I will investigate whchi test fails due to Keycloak and why.
